design/deploy-steps.md

```diff
@@ -0,0 +1,134 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# Customizable deployment procedure
+
+## Status
+
+One of: implementable
+
+## Summary
+
+This design proposes customizing the deployment procedure by providing a high
+level abstraction around Ironic [deploy steps][deploy-steps], as well as
+allowing per-host deployment (IPA) images.
+
+[deploy-steps]: https://docs.openstack.org/ironic/latest/admin/node-deployment.html#deploy-steps
+
+## Motivation
+
+By default the Ironic agent downloads a provided image via HTTP(s), converts it
+(if needed) and writes directly to the disk. This procedure works for any
+distributions, but may not suit some cases well. A prominent example is CoreOS,
+which has its own installer (`coreos-installer`) that allows more
+customization.
+
+### Goals
+
+- A way to execute custom actions during deployment.
+- A way to replace the default deploy procedure with the custom actions.
+- A way to provide a link to the Ironic agent image per host.
+
+### Non-Goals
+
+- Expose the full functionality of Ironic deploy steps.
+- Building custom agent images.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a CoreOS user I want to use `coreos-installer` to write the CoreOS image
+without losing any functionality that Ironic offers (unlike the live ISO
+workflow).
+
+#### Story 2
+
+As an operator I want to execute site-specific actions during deployment. I'm
+ready to provide them as plugins to the Ironic agent inside my custom agent
+image.
+
+## Design Details
+
+Additions to `BareMetalHost`:
+
+- New field `AgentImageURL`. It will accept an optional link to a custom agent
+  ISO image.
+
+NOTE: custom kernel/initramfs are out of scope for this proposal.
+
+- New field `DeploySteps` with a list of deploy steps to apply. Each step
+  is a structure with fields `Name` (string) and `Args` (JSON).
```

Comment:
One of the arguments we will need to provide is the network data. Is that merged with the map from the args parameter or do we expect the user to provide it in that field?

Reply:
If you need to pass explicit network data to `coreos-installer`, it will be one of the arguments, like in my prototype: https://github.com/dtantsur/ironic-agent-image/blob/main/hardware_manager/ironic_coreos_install.py#L24

Reply:
That means the user needs to provide the data 2 places. I suppose it may need to be in 2 formats? The existing field is a reference to a Secret resource, but posting it inline in the step would mean (a) adding potentially large data to the host resource and (b) it couldn't be protected.

What if instead of a JSON blob field, we ask for a Secret reference, and pass the keys and values from that Secret through? That preserves the security, lets us use a type that accepts arbitrary data in key/value pairs, and keeps the host resource itself small.

Reply:
It would also be useful to look at what the template controller generates. Maybe we want to use a reference to one of those resources instead, to tie in with the templating capabilities.

Reply:
> That means the user needs to provide the data 2 places.

I'm not sure I'm following. With this approach there is only one place in BMO: this one. BMH.NetworkData will be unused.

> What if instead of a JSON blob field, we ask for a Secret reference

What you're suggestion requires explicit knowledge of CoreOS. We've agreed not to do that, thus we cannot offer special handling of any arguments.

Or have I misunderstood the proposal?

Reply:
> > That means the user needs to provide the data 2 places.
> 
> I'm not sure I'm following. With this approach there is only one place in BMO: this one. BMH.NetworkData will be unused.

I guess the user will have to know not to use the networkData field when their custom steps need that information passed directly.

It feels like it would be cleaner if we defined an API for the step arguments that we will pass to the backend. For example, if we said we were going to assemble a standard set of data to pass to the steps, including user data, network data, etc., and that the step-specific arguments would be added to that payload then all steps would receive the same info and could ignore what they don't need/want.

> > What if instead of a JSON blob field, we ask for a Secret reference
> 
> What you're suggestion requires explicit knowledge of CoreOS. We've agreed not to do that, thus we cannot offer special handling of any arguments.
> 
> Or have I misunderstood the proposal?

I expect the user to need to understand how the custom IPA image they're using works. What I don't want is to have the BMO need to understand it.

I'm suggesting that we follow the pattern established elsewhere in the API for handling "a large blob of data we can't validate" and put it into a Secret, which gives us a key/value pair data type that has no expectations for validation and that kubernetes users are used to manipulating. We could also use a ConfigMap, but I think we're using Secret everywhere else and consistency is nice.

If we combine the 2 ideas in this thread, then we could say that all steps will receive a payload with a `networkData` key containing the content of the `networkData` secret, a `userData` key containing the content of the `userData` secret, and a `stepData` key containing the step-specific data from that secret. We could also include a `hostData` key with some additional info, like the host resource name or other values that might be useful during provisioning.

Reply:
> > could ignore what they don't need/want
> 
> Ironic steps do have validation, you cannot pass arbitrary data to them.

What defines the validation?

> > I'm suggesting that we follow the pattern established elsewhere in the API for handling "a large blob of data we can't validate" and put it into a Secret, which gives us a key/value pair data type that has no expectations for validation and that kubernetes users are used to manipulating.
> 
> Ah, so pass all arguments as one secret?

Yes, one per step.

> > all steps will receive a payload with a networkData key containing the content of the networkData secret, a userData key containing the content of the userData secret, and a stepData key containing the step-specific data from that secret
> 
> Will work nicely for us, but see above. This will limit this proposal to our use case.

I'm not sure why you see it as limited. It would use a bunch of details from other fields on the host as well as the step-specific arguments to construct a well-defined payload. The user would have complete control over what goes into the payload, but wouldn't have to construct the full payload themselves. Step authors would know what to expect to receive as arguments. Where is the missing flexibility?

Reply:
> What defines the validation?

The step itself. Like here https://github.com/dtantsur/ironic-agent-image/blob/main/hardware_manager/ironic_coreos_install.py#L23-L48

> I'm not sure why you see it as limited.

We won't be able to execute any steps that are not written specifically for metal3. E.g. ironic ships a step to inject files into the instance, we won't able to use it, it accepts a different set of arguments.

> The user would have complete control over what goes into the payload, but wouldn't have to construct the full payload themselves. 

But there will be network_data/user_data/step_data arguments, no? Or have I misunderstood it again?

Reply:
> > I'm not sure why you see it as limited.
> 
> We won't be able to execute any steps that are not written specifically for metal3. E.g. ironic ships a step to inject files into the instance, we won't able to use it, it accepts a different set of arguments.

The step-wise validation was the piece I was missing to understand this. Thanks for explaining that. I guess users are going to have to understand which steps are available and what arguments those steps take.

Aside from the arguments passed to the steps, how much do we think the step list will vary between hosts? Would it be useful to define an API with a set of steps that the BMO could use as a template to build the instructions to give to Ironic?

Reply:
> how much do we think the step list will vary between hosts?

Steps come from two sources: the driver and the agent. The former can vary quite a bit, the latter can in theory, but very rarely in practice.

> Would it be useful to define an API with a set of steps

You mean, in Ironic? Would love to, but there is a huge chicken-and-egg problem with booting the agent. We don't know which steps are in the agent until we boot it. 

Now, in our situation it's fine because we boot the agent early for inspection/cleaning (unless these are disabled). But in a general Ironic context it's normal when the agent is only boot for deployment itself, too late to return anything to a user. For this reason we haven't implemented any "get clean/deploy steps" API.

Reply:
> > Would it be useful to define an API with a set of steps
> 
> You mean, in Ironic? Would love to, but there is a huge chicken-and-egg problem with booting the agent. We don't know which steps are in the agent until we boot it.
> 
> Now, in our situation it's fine because we boot the agent early for inspection/cleaning (unless these are disabled). But in a general Ironic context it's normal when the agent is only boot for deployment itself, too late to return anything to a user. For this reason we haven't implemented any "get clean/deploy steps" API.

Well, I meant here in metal3. So that instead of listing the steps to provision each host individually on that host, we would have a separate API resource that lists the steps and the host would have a reference to that resource. That does make passing per-host settings a bit more complex, though, so maybe that's not ideal.

Reply:
> separate API resource that lists the steps

Where will this information come from given that we cannot ask Ironic for that?